### ISSUE TITLE
Support MAME's Compressed Hunks of Data (CHD) filesystem format

### ISSUE THREAD

**Issue by MartyShepard** [dosbox-staging/dosbox-staging#166]
At Github there is a stand-alone version of MAME CHD for the formats v1-5
https://github.com/rtissera/libchdr

https://github.com/mamedev/mame/tree/master/src/lib/util  (Look for CHD Files at top)
https://github.com/stenzek/duckstation/blob/master/src/common/cd_image_chd.cpp
https://github.com/hayate891/nulldcr/blob/master/plugins/ImgReader/chd.cpp
https://github.com/FCare/Kronos/tree/extui-align/yabause/src/libchdr

There are already a lot of emulators that read compressed CD-Roms (ISO, BIN, CUE with Audio) in CHD format Container and the use the standalone version.

I tried it myself but my skill in C ++ is not really good enough for that. I use and code only in MingW on Windows without Debug Features like Visual Studio. I begin in cdr_image.cpp for a LoadCHD and give up with Readsektor in Loadimage Routine (drive_iso.cpp) . :(

It would be great if Dosbox could use the compressed CHD format for CDRoms.
Overall, it would be an advantage since there is finally a standard for compressed CD-ROMs.

regards

**Comment by krcroft** [comment]
@MartyShepard , thank you for the suggestion and looking at how to include this.

There is collective interest in achieving "_DOS games-as-ROMs_": a tidy and immutable single-file that bundles up a game's installation directory and dependencies (such as CDROM) that can be runnable by DOSBox and other emulators.  

We've discussed this in #99, and @dreamer and I have chatted offline about this too.  

Like CHD, the desired format will be a compressed filesystem inside of which sits the game's files and directories, plus a common set of reserved "domain-specific" content.  We haven't defined this yet, but here's an example of this reserved & defined content:

-  **Extensible metadata** surrounding the game: name of the game, release date, game developer, publisher, screenshots, package materials (front/back box art), documentation and quick reference sheets, SHA checksums, and so on. This area will be extensible and versioned, as the initial content might be thin, but allows the community to incrementally grow the medata inside each game. 
- **Virtual hardware record** used to configure the emulator running the _as-bundled game_, such as CPU type and speed, video mode, sound-card and configuration, MIDI type and configuration, mixer volume adjustments, joystick support, and so on.
- **Configuration variations**, this would be one or more sets of game-specific config files, each paired with a virtual hardware record, allowing the game to be launched under various supported configurations.  For example: EGA or VGA; Adlib, SB16, or MT32.
- **CDROM** data and audio tracks in FLAC or Opus formats (or future supported best-in-breed formats).

CHD version 5 is certainly a solid option for this, however per-block LZMA decompression can add significant latency (for example, using CHDs with PS1 emulation on an RPi3 requires the smallest CHD block size because at the larger end the decompression latency is too slow that it causes frame and audio stuttering). 

To minimize block decompression latency, CHD should use ZStandard compression, which is typically 10-fold (or more) quicker than LZMA at decompression. We'd also want to it to support zero-compression blocks, so our Opus and FLAC data aren't double-compressed (we already take a latency hit decoding those audio blocks).   This might all be early optimization, but the benefits are documented and measurable, and in theory this should be pretty easy to add to CHD (if we decide that CHD is the best fit).

**Comment by i30817** [comment]
CHD already has a versatile metadata facility, but i agree about the double compression gotcha.

However, i'm not quite sure it isn't already done. CHD is supposed to be 'non-lossy' (on all cases, hd or cdrom) and thus it uses flac for audio tracks iirc (unless the simulationist faction of MAME managed to ruin that already).

Introducing a lossy audio compression just for support for GOG images is the kind of thing that i can see a dogmatic upstream go 'nope' about.

**Comment by i30817** [comment]
Also there is this issue that you might interested in for CHD:

https://github.com/libretro/beetle-psx-libretro/issues/587

tl;dr : chd parent-child compression support doesn't exist in libchd yet and it requires some weird handling on the emulators because the chd v5 header doesn't store a path to the 'parent' but a sha1sum, so if you want to use it, you have to have a 'scan step' or a convention to find the parent.

**Comment by krcroft** [comment]
I'm not seeing any API calls that let CHD provide us with a transparent internal filesystem-like API. So this means we need to provide it with our own filesystem (an ISO or VFAT blob), and then when we seek and read chunks out of the CHD, feed those into our ISO or FAT filesystem parser code, which in turns figures out that those chunks representative file1, file2, and so on.  

Even though it's doable, it feels excessively complex. We have no need to feed our emulator with some custom ROM or binary structure that Sony or Nintendo cooked up (in which case, CHD makes perfect sense). 
Instead, DOS games simply want to operate on files directly: fopen(), fread(), and fseek().  These calls, in the case of DOSBox's  `mount` command, simply translate through to similar standard C and C++ calls on the local filesystem. 

So I was hoping we would have a library that lets us perform those filesystem calls on it, except they hit the compressed files in the archive, as opposed to coming off the local filesystem.

**Comment by i30817** [comment]
Does chd 'internal' support copy on write for hdds? It's kind of the 'feature' that can't be abstracted by dosbox windows support without wasting a lot of space on the first copy.

**Comment by i30817** [comment]
I feel the last post wasn't clear enough even if it's informative, so tl;dr: I suspect that if CHD parent-child relationships are supported for writable hhds chds, you could get 'copy-on-write' for free by making a 'empty' child to receive the writes when you mount the 'child'.

**Comment by krcroft** [comment]
Yeah; CHD gives back read-only blocks. No way to write into it, which we expect to be true for whatever format we choose. 

So for writes, we would use DOSBox's recently added overlay feature and place any written files outside of the archive in a 'sidecar' directory or like you're thinking, into a writable archive (say a zip file), which would be nice and tidy. 

- game name.rom
- game name.sav

**Comment by i30817** [comment]
If it *never* writes, it must already be writing somewhere else at least in MAME, because the MAME set has chds that are hard drives. It probably uses the sha1sum as a id of the 'file' (or dir) it writes to.

I believe using the sha1sum as a id is better than anything using the name, even if becomes 'hard' to delete manually, it will never cause (non-malicious) problems with false positives.

**Comment by dreamer** [comment]
I don't know CHD format, so can't comment on it, but I can share some additional use cases or features we want or need to support, or even started working on:
- As @krcroft mentioned, copy-on-write using existing DOSBox' mount overlay mechanism; this has nothing to do with checksumming in my opinion. ATM I think we should have a mechanism like this: special metadata field declaring application name / game name - having the same role as GOG "slug" names - used for identification of a program in filesystem (e.g. under XDG hierarchy) - so instead of having single .conf file shared between applications (`~/.dosbox/dosbox-<suffix>.conf` right now), each "ROM" DOS game, would get it's own hierarchy: e.g. `~/.config/dosbox/<game-name>/` to store user-changed conf options, per-game mapper files, AND game overlay directory (for storing game saves).
- to get above feature working, we need a way to declare per-game filters, so game packager could filter files, that are incorrectly flagged with write access by DOS program - initial work in this area is in progress in #143 
- checksumming *might* be beneficial, but I'm not sure is necessary except for verification of internal ROM integrity; whatever checksum we'll use, it won't be sha1. We could use similar object structure as Git is using for object files and pack files, or do something simpler, like using tar + checksums (however modern tar implementations have irritating behaviour of creating non-reproducible files by default due embedding SELinux attributes and such).
- We can't think about technicalities and users of ROM files only, but also about packagers - for this reason, I think the tools for creating "DOS-ROMS" should consist of small, specialized tools (unix-style small tools, not a set of additional switches added to dosbox-staging).
- Yes, we want ability to embed uncompressed files for FLAC files and other, but ALSO compress raw images - besides gamers, important target user group is "game archivists", and they might have very valid reasons to store raw CD images.
- Mounting and switching between CD images (in whatever format) needs to be possible as a "lazy" operation - if the game consists of 7CDs, we don't want to load them all at once in memory.

*Whatever I wrote are only suggestions, point out whenever I say something stupid or there is a better technical solution possible ;) I think after few more comments we could start working on RFC/design doc*

**Comment by i30817** [comment]
The sha1sum i'm mentioning is already part of the chd format in the header. When you create a chd in chdman, you already have it fixed 'forever' so you're checksumming per-se, 'just' using a cached unique id that refers to that single file.

Using anything else to map which 'written data' applies over the 'original data' is a waste of time imo.

**Comment by dreamer** [comment]
@i30817 
> Using anything else to map which 'written data' applies over the 'original data' is a waste of time imo, and may cause false positives / bugs if the 'mapping' is done with hdd image that can change without the id changing. Other metadata like the game name can exist, sure, but the sha1sum to map from the 'original' to the 'modifications' seems like straightforward foolproof solution.

I don't understand what you mean. Obviously we're not going to support writing to CHD image / "DOS-ROM" format, it will be read-only way of distributing DOS software. What do you understand by "modifications"?

**Comment by i30817** [comment]
When i wrote about writing/modifications i'm thinking of the support for persistent modifications that the CHD format - as a hard drive image format - should support. CHDs are not written to directly (i learned here) but that doesn't mean modification don't happen or disappear every run - they're just saved on a different file already.

Otherwise MAME hd image in the MAME dump set couldn't work.

**Comment by dreamer** [comment]
Because @MartyShepard mentioned it specifically in the context of CD images, not as generic HDD image format. I see no reason to support HDD image format with data compression supporting reads and writes (like HDD images in VirtualBox), but I see plenty of reasons to support read-only distribution format (like game ROMs, but for DOS software).

**Comment by MartyShepard** [comment]
Basically I just want to read the CHD (CDRoms) in Dosbox no more.

For me, the first purpose is to compress all the DOS CDRom Isos that i have created from my CDs, to use them "at the same time" and to save "hd space".

I have many games for Dosbox that use more than 3 CD-ROMs and data CD's can save a lot of space instead of pure (ISO, BIN) on the disc. But also a collection of CDRom games that only use 1 CDRom can save a lot of space. I see that from my PSX and Dreamcast Collection, I use CHD tool from Mame and it goes really well and play very good.

Under no circumstances should it be expanded so heavy as in Mame and add the CHD-HD format  into Dosbox. It will be far too complex. 

Basically, the user should decide whether to use the CHD format or not. For me it would be perfect if it found support and would be overjoyeds.

regards

**Comment by dreamer** [comment]
Clarified, thanks :)

So basically we were all talking each about different usecases, that are a bit similar, but not the same.

This issue will continue to be for discussion and implementation of new usecase for imgmount:

> Z:\> imgmount D: file.chd -t cdrom -fs chd

@i30817 was talking about RW mountable format. I suspect as an alternative for #99, right?

me and @krcroft were talking about distribution format for pre-installed games, with possibility of embedding CD images inside (so another usecase, maybe CHD format applies for this usecase, maybe not) - we don't have an issue for this feature (yet).

**Comment by i30817** [comment]
> @i30817 was talking about RW mountable format. I suspect as an alternative for #99, right?

No - they're different things and chd wouldn't be useful for that usecase, simply because it's opaque to the native filesystem.

I *was* talking about it as a 'distribution format for pre-installed games' (though i think 'embedding CD images inside' is not a good idea). I simply think a  'distribution format for pre-installed games'  *needs* to be mountable for the game to run without additional steps; which means it needs to be RW, which in CHD means the writes need to be redirected to another file.

I have no idea how this is considered controversial.

**Comment by krcroft** [comment]
> a 'distribution format for pre-installed games' needs to be mountable (...)

Yes.

> ... for the game to run without additional steps; which means it needs to be RW

Games that never open their files with write flags (like the old 80s games that never saved settings, high scores, or supported save-games). Their files can be marked read-only, the filesystem can be read-only, and the block layer can be read-only.

On the other hand, games that _do_ use the write flag but never actually write a single byte need this change ( https://github.com/dreamer/dosbox-staging/pull/143)  run within an immutable environment. 

And finally, games that both use the write flag and write actual bytes need a bonafide writable filesystem, so we need that outlet somewhere. 

It certainly won't be allowed to write into the single-file  ROM package, even if the format does support writes. We want the ROM to be immutable, pristine, with verifiable integrity - such that the ROM can represent the one and only clean-room installation of a specific version of a game.  That leaves all all users writes landing somewhere external to the ROM, in either a writable container file or directory structure, as organized as we can make it.

**Comment by MartyShepard** [comment]
Here is a screenshot of what I mean. That is my try. This is an attempt but is only recognized as Audio_CD as I am without a plan in the read sector routine what i have to do.

![Aufnhame00033](https://user-images.githubusercontent.com/8560193/74091805-3dd68f00-4abc-11ea-9e35-b7c26df88e61.png)

regards

**Comment by theMK2k** [comment]
Talking about **metadata**:

With [Metropolis Launcher](https://metropolis-launcher.net/) I ship **moby.db** which is an extensive SQLite database dump of what mobygames.com (the imho best source of metadata for DOS games) offers.

Feel free to use it as a metadata source of your CHD-based ROM format. With this, you won't start thin ;)

**Comment by rtissera** [comment]
libchdr https://github.com/rtissera/libchdr is being overhauled at the moment so let me know if you want to discuss CHD support which would be amazing.
Regarding the LZMA decompression speed concerns and induced latency, I am considering upgrading LZMA routines from 16.04 to 19 and possible AArch32 / AArch64 optimizations even if GCC 9 and GCC 10 already auto-vectorize code using NEON.

**Comment by nemo93** [comment]
> Basically I just want to read the CHD (CDRoms) in Dosbox no more.
> 
> For me, the first purpose is to compress all the DOS CDRom Isos that i have created from my CDs, to use them "at the same time" and to save "hd space".
> 
> I have many games for Dosbox that use more than 3 CD-ROMs and data CD's can save a lot of space instead of pure (ISO, BIN) on the disc. But also a collection of CDRom games that only use 1 CDRom can save a lot of space. I see that from my PSX and Dreamcast Collection, I use CHD tool from Mame and it goes really well and play very good.
> 
> Under no circumstances should it be expanded so heavy as in Mame and add the CHD-HD format into Dosbox. It will be far too complex.
> 
> Basically, the user should decide whether to use the CHD format or not. For me it would be perfect if it found support and would be overjoyeds.

=> I do fully agree with @MartyShepard here and I'd love to see CHD support in Staging. That would really help those with quite big CD library as the gain in term of storage space is impressive. It's also a bit easier to manage (no more iso, cue or img to deal with). I'd say we're talking 30-40% of saving (bin/cue vs chd) based on my personal experience. That would also be a nice differentiator compared to other forks. 

I just know implementation is not that straightforward (I recall some chaps on libretro asking for assistance with some cores). Perhaps it's best to leave it for now and reopen at a better time?

**Comment by kcgen** [comment]
I'll start by saying I'm a huge fan of CHD: I used it for my own TG16, SegaCD, and PSX sets on the Pi, and have scripts to assist in the conversion from clean dumps. 

In the time prior to working on Staging, I too wanted CHD/Zip/PhysFS for DOSBox; but having more control now, I want something that would check more of my boxes: https://github.com/dosbox-staging/dosbox-staging/wiki/DOS-ROM-Format.  

Why not merge Zip/CHD support now - it's still better than nothing!  From a maintainers perspective, we're talking about truck loads of code that permeates all IO areas in very non-elegant ways. Doing so would kill the more grand vision without having checked most boxes.

So I agree that it's better than nothing, but the feature compromises are far too great for me as a maintainer to outweigh the benefits, when we have plans of something (possibly) better.

**Comment by i30817** [comment]
@kcgen your plan doesn't appear to mention support of softpatching, which chd has.

Personally i also want #99 for the reasons i explain there though it's a longshot.

**Comment by kcgen** [comment]
hi @i30817 , I see the #99 as a do-able extension of `imgmount`'s existing functionality that won't preclude implementing a packed-ROM format (or vice-versa).

**Comment by i30817** [comment]
Ok that's good, just checking. Sometimes people monofocus on a single format to solve it all and a packed rom format would never solve the case i want #99 for.

No chance of using libchd and extend it into that rom format you want? Because chd has a metadata header you can put anything in and already has the capability to softpach.

**Comment by kcgen** [comment]
> No chance of using libchd and extend it into that rom format you want

I'm positive with enough effort and ideas like this, CHD could hold what we need; however it would fail to meet many other requirements; so it's already a failed compromise out of the gate. 

For example: it still doesn't use zstandard. 

 It would still impose a much greater burden on the maintenance side than I can stomach. 

I'm also worried that stuffing custom blobs in the metadata would not be extractable using existing 3rd party tools (which we have a requirement for). 

We'll see what we can accomplished; if we can't get there then so be it - in that case, we will be thankful to have CHD or Zip to backstop this effort.

**Comment by i30817** [comment]
Ugh, i'll be honest, a file format for game packing that just uses zip+some file inside of it as a 'header' is something that is rather inappropriate for larger games and hard-drive images (which is the whole reason that formats like qcow and chd were created).

If you're going to try to use zlib type of formats just because 'unrar' or zip can open them, you're taking on the massive disadvantage that zip was and is (except for some variants that are just as incompatible) a fileformat for archiving, not read only compression, and thus when you want to seek backwards something inside of a big file, you have no alternative but decompressing that big file from the start.

**Comment by kcgen** [comment]
Right; we're not talking about using an archive format - for the reason you're suggesting 👍  the leading off-the-shelf format is squashfs, which has a portable client library, user-space tools, is fast, has pluggable compression formats, and is extremely well maintained and tested.

**Comment by i30817** [comment]
Quashfs is cool but it's not patchable. I think.

Anyway, i realized something meanwhile. It would be 'good for you' if you get get a fileformat that is 'easy to use' on a in-memory filesystem. Specifically a overlay filesystem, which you're going to need anyway if you plan to make your format read only. This is a very similar problem to softpatching btw.

The reason is that the main problem with a apparently unrelated functionality - savestates - is that I/O being outside of what the emulator records breaks savestates. 

You can't have a emulator writing to a file, suspend that in a savestate, let it finish writing, do something else that writes to that file, then restore the savestate and restart writing to the file without that breaking. 

However, if you could _record_ all the changes in the timeline 'until the savestate' including any in between I/O and store that in savestate and if you could glue that to a 'pristine' image of the hard-drive/game at the start and resume those partially done I/O, it would work. And you can probably do it with a overlay filesystem (i suspect it would be tough).


They're very similar problems to softpatching if you think of them this way.

**Comment by kcgen** [comment]
@i30817 , here's our ongoing document of what we want: https://github.com/dosbox-staging/dosbox-staging/wiki/DOS-ROM-Format  ; I realize now that you might not have seen it yet.

**Comment by i30817** [comment]
I mean, i'm not the programmer here. I was just mentioning the '3 related things' because it seemed you had overlooked the fact that savestates design would be helped with a overlay filesystem - preferably one that allows multiple levels of overlay like chd does to also allow softpatching at the same time.

**Comment by kcgen** [comment]
>  it seemed you had overlooked the fact that savestates design would be helped with a overlay filesystem

Yes, it's fully understood that games expect to write data, and those writes will need to go somewhere writable (given we're dealing with an immutable read-only ROM format). DOBox's existing overlay system will be perfect for the job (and that's the plan).

**Comment by Grounded0** [comment]
Changed closed type to not completed.

**Comment by Grounded0** [comment]
Changed closed type to not completed.
